### ISSUE TITLE
[Bug]: ValueError: Chunk shape must not be greater than data shape in any dimension. (512, 512) is not compatible with (1128, 409)

### ISSUE THREAD

**Issue by gjustin40** [mobigroup/gmtsar#63]
I have already looked at the issue #40, but it doesn't seem to align with my situation, so I have a question.

I am currently trying the example from the blog post, specifically **[Intf.ipynb.](https://www.patreon.com/posts/pygmtsar-python-82433177)**

The code I used is as follows:
```
RESOLUTION   = 90
sbas.set_pins([137.381543, 34.419120, 137.85567, 35.001281])
# The error below occurred when I used the entire range without setting
sbas.set_pins()

# miss optional "func" argument when post-processing is not required
decimator = sbas.pixel_decimator(resolution_meters=RESOLUTION, debug=True)

# default parameters: wavelength=200, psize=32, func=None (no postprocessing)
#, wavelength=200, func=decimator
sbas.intf_parallel(pairs, wavelength=400, func=decimator, n_jobs=-1)
```

The error occurred as follows:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[25], line 3
      1 # default parameters: wavelength=200, psize=32, func=None (no postprocessing)
      2 #, wavelength=200, func=decimator
----> 3 sbas.intf_parallel(pairs, wavelength=400, func=decimator, n_jobs=-1)

File /opt/conda/lib/python3.10/site-packages/pygmtsar/SBAS_intf.py:40, in SBAS_intf.intf_parallel(self, pairs, n_jobs, **kwargs)
     38 # this way does not work properly for long interferogram series
     39 with self.tqdm_joblib(tqdm(desc='Interferograms', total=len(pairs)*len(subswaths))) as progress_bar:
---> 40     joblib.Parallel(n_jobs=n_jobs)(joblib.delayed(self.intf)(subswath, pair, **kwargs) \
     41         for subswath in subswaths for pair in pairs)

File /opt/conda/lib/python3.10/site-packages/joblib/parallel.py:1098, in Parallel.__call__(self, iterable)
   1095     self._iterating = False
   1097 with self._backend.retrieval_context():
-> 1098     self.retrieve()
   1099 # Make sure that we get a last message telling us we are done
   1100 elapsed_time = time.time() - self._start_time

File /opt/conda/lib/python3.10/site-packages/joblib/parallel.py:975, in Parallel.retrieve(self)
    973 try:
    974     if getattr(self._backend, 'supports_timeout', False):
--> 975         self._output.extend(job.get(timeout=self.timeout))
    976     else:
    977         self._output.extend(job.get())

File /opt/conda/lib/python3.10/site-packages/joblib/_parallel_backends.py:567, in LokyBackend.wrap_future_result(future, timeout)
    564 """Wrapper for Future.result to implement the same behaviour as
    565 AsyncResults.get from multiprocessing."""
    566 try:
--> 567     return future.result(timeout=timeout)
    568 except CfTimeoutError as e:
    569     raise TimeoutError from e

File /opt/conda/lib/python3.10/concurrent/futures/_base.py:458, in Future.result(self, timeout)
    456     raise CancelledError()
    457 elif self._state == FINISHED:
--> 458     return self.__get_result()
    459 else:
    460     raise TimeoutError()

File /opt/conda/lib/python3.10/concurrent/futures/_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

ValueError: Chunk shape must not be greater than data shape in any dimension. (512, 512) is not compatible with (1128, 409)
```

I thought the error might have been due to the range set for `set_pins` being too small, so I tried increasing the spin range. Furthermore, I even attempted running it without setting spin, running it over the entire range, but the error remains the same, with only the numbers in (xx, yy) changing.

What is the reason for this error, and what should I do to resolve it?
(By the way, i'm using docker that you deployed it)
Thanks

**Comment by AlexeyPechnikov** [comment]
PyGMTSAR uses chunk-based processing, where large rasters are divided into smaller square tiles (512 x 512 pixels by default) to significantly reduce memory consumption. Larger chunks result in faster processing but require more memory, while smaller chunks process more slowly but demand less memory. If you encounter an error because your data grids are smaller than the default chunk size in one dimension, you can either increase the data file sizes by processing a larger area or reduce the chunk size. To decrease the chunk size, do so right after loading the PyGMTSAR library and before any processing:
```
from pygmtsar import datagrid
datagrid.chunksize = 256
```
By the way, the development version of PyGMTSAR can automatically adjust the chunk size to prevent this error.

**Comment by gjustin40** [comment]
Thank you for your response. 
I will share the results as soon as they become available.

Regarding your information about the development version, 
I created a new Docker container to run pygmtsar because I didn't want to use Jupyter Lab.
(Of course, I used Docker you provided for the above question)

However, although everything went well with the import, I encountered the following error in the prepare.ipynb at the `sbas.reframe_parallel()` part. 

```
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/conda/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py", line 463, in _process_worker
    r = call_item()
        ^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/joblib/externals/loky/process_executor.py", line 291, in __call__
    return self.fn(*self.args, **self.kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/joblib/parallel.py", line 589, in __call__
    return [func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/joblib/parallel.py", line 589, in <listcomp>
    return [func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.11/site-packages/pygmtsar/SBAS_reframe.py", line 154, in reframe
    tmpazi = prm.SAT_llt2rat([pins[0], pins[1], 0], precise=1, debug=debug)[1]
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: index 1 is out of bounds for axis 0 with size 0
"""

The above exception was the direct cause of the following exception:

IndexError                                Traceback (most recent call last)
Cell In[46], line 1
----> 1 sbas.reframe_parallel()

File /opt/conda/lib/python3.11/site-packages/pygmtsar/SBAS_reframe.py:219, in SBAS_reframe.reframe_parallel(self, dates, n_jobs, **kwargs)
    217 # process all the scenes
    218 with self.tqdm_joblib(tqdm(desc='Reframing', total=len(dates)*len(subswaths))) as progress_bar:
--> 219     records = joblib.Parallel(n_jobs=n_jobs)(joblib.delayed(self.reframe)(subswath, date, **kwargs) \
    220                                              for date in dates for subswath in subswaths)
    222 self.df = pd.concat(records)

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:1952, in Parallel.__call__(self, iterable)
   1946 # The first item from the output is blank, but it makes the interpreter
   1947 # progress until it enters the Try/Except block of the generator and
   1948 # reach the first `yield` statement. This starts the aynchronous
   1949 # dispatch of the tasks to the workers.
   1950 next(output)
-> 1952 return output if self.return_generator else list(output)

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:1595, in Parallel._get_outputs(self, iterator, pre_dispatch)
   1592     yield
   1594     with self._backend.retrieval_context():
-> 1595         yield from self._retrieve()
   1597 except GeneratorExit:
   1598     # The generator has been garbage collected before being fully
   1599     # consumed. This aborts the remaining tasks if possible and warn
   1600     # the user if necessary.
   1601     self._exception = True

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:1699, in Parallel._retrieve(self)
   1692 while self._wait_retrieval():
   1693 
   1694     # If the callback thread of a worker has signaled that its task
   1695     # triggered an exception, or if the retrieval loop has raised an
   1696     # exception (e.g. `GeneratorExit`), exit the loop and surface the
   1697     # worker traceback.
   1698     if self._aborting:
-> 1699         self._raise_error_fast()
   1700         break
   1702     # If the next job is not ready for retrieval yet, we just wait for
   1703     # async callbacks to progress.

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:1734, in Parallel._raise_error_fast(self)
   1730 # If this error job exists, immediatly raise the error by
   1731 # calling get_result. This job might not exists if abort has been
   1732 # called directly or if the generator is gc'ed.
   1733 if error_job is not None:
-> 1734     error_job.get_result(self.timeout)

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:736, in BatchCompletionCallBack.get_result(self, timeout)
    730 backend = self.parallel._backend
    732 if backend.supports_retrieve_callback:
    733     # We assume that the result has already been retrieved by the
    734     # callback thread, and is stored internally. It's just waiting to
    735     # be returned.
--> 736     return self._return_or_raise()
    738 # For other backends, the main thread needs to run the retrieval step.
    739 try:

File /opt/conda/lib/python3.11/site-packages/joblib/parallel.py:754, in BatchCompletionCallBack._return_or_raise(self)
    752 try:
    753     if self.status == TASK_ERROR:
--> 754         raise self._result
    755     return self._result
    756 finally:

IndexError: index 1 is out of bounds for axis 0 with size 0
```

It seems to be an issue related to the version of a specific module,
Can you provide any ideas for solving this issue?

Thanks

**Comment by AlexeyPechnikov** [comment]
The last error looks as a result of incorrect DEM file. Do you use PyGMTSAR DEM downloading function?

Development version is not ready to use because I’m maintaining complete code refactoring to realize the new announced features.

**Comment by gjustin40** [comment]
I am trying to apply the method you mentioned.

```
from pygmtsar import datagrid
datagrid.chunksize = 256
```

Currently, The situation is like below
![image](https://github.com/mobigroup/gmtsar/assets/25999141/a82ef5c3-2624-4c3c-88e9-473dc0a1b2bc)

When I ran the code, initially all the CPUs were at 100% usage, and I could see commands like 'diff_phase' being executed in htop. So, in the beginning, I could tell that it was working fine. 
However, after the progress bar filled up once, when I checked htop, it seemed like there were no functions being executed.
(I filtered by python in htop)
![image](https://github.com/mobigroup/gmtsar/assets/25999141/34f2fdc4-12f5-4440-9d81-7844f1e418fe)


I've encountered a similar issue before, and I thought it was just a computationally intensive task, so I waited for about 3-4 days, but there was no change. So, I thought something might be stuck or a bug had occurred.

I'm worried that the same issue might happen today, and I'm attaching the current htop results as well. 
Is it in a working state?

server spec
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          40
On-line CPU(s) list:             0-39
Thread(s) per core:              2
Core(s) per socket:              10
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Silver 4210R CPU @ 2.40GHz
Stepping:                        7
CPU MHz:                         1000.065
CPU max MHz:                     3200.0000
CPU min MHz:                     1000.0000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
L1d cache:                       640 KiB
L1i cache:                       640 KiB
L2 cache:                        20 MiB
L3 cache:                        27.5 MiB
NUMA node0 CPU(s):               0-9,20-29
NUMA node1 CPU(s):               10-19,30-39



Thanks

**Comment by AlexeyPechnikov** [comment]
Yes, small chunk processing is not stable often. In the PyGMTSAR example S1A_Stack_CPGF_T173.ipynb, 60m resolution processing is used to ensure the output grids have a size larger than the default chunk size. I see you changed the resolution (decimation=6), so reverting the resolution to its default value of 90m (decimation=4) should be enough to fix the issue.

Also, you can update the PyGMTSAR library in the Docker container by running the code below in any notebook cell. After updating, make sure to restart the Jupyter kernel and re-run the notebook.
```
#!{sys.executable} -m pip install pygmtsar >/dev/null
!{sys.executable} -m pip install -U pygmtsar 
```

P.S. The "top" utility is very outdated, and it's recommended to use "htop" instead to view your processes, CPU cores, RAM usage, and more.

**Comment by gjustin40** [comment]
I'm still stuck at the `sbas.intf_parallel` part.

1. resolution: 90m
2. bbox: lat(34.7241 ~ 34.7479), lon(137.5956, 137.6078)
3. datagrid.chunksize = 128
4. SLC Data: 5 interferogram stacks


I've monitored using htop.
Initially, when I run `sbas.intf_parallel`, it uses 100% of all CPUs for computation, but at some point, it continues in a state where the process named `/opt/conda/bin/python -c from multiprocessing.spawn import spawn_main; spawn_main(tracker=fd=103, pipe_handle=166) --multiprocessing` occupies 4~5% of the CPU, as shown in the htop screen I mentioned above.

On the Jupyter notebook, the progress bar isn't increasing, and it's been going on for several days without finishing. Thus, I can't determine whether there's an error or a deadlock.

Initially, I tried with 31 SLC data sets, but when it didn't finish for a long time, I reduced it to 5. Yet, it still remains in an unfinished state.

All I can share about the current situation is that `sbas.intf_parallel` is operating on the Jupyter notebook.

What could be the problem?
Thank you.
